storage/rocksdb/ha_rocksdb.cc

```diff
@@ -15076,6 +15094,513 @@ void rdb_tx_multi_get(Rdb_transaction *tx,
+
+  tx->multi_get(m_pk_descr->get_cf(), mrr_n_elements, mrr_keys, mrr_values,
+                mrr_statuses, mrr_sorted_mode);
+
+  return 0;
+}
+
+void ha_rocksdb::mrr_free() {
+  // Free everything
+  if (mrr_enabled_keyread) {
+    m_keyread_only = false;
+    mrr_enabled_keyread = false;
+  }
+  mrr_free_rows();
+  delete mrr_rowid_reader;
+  mrr_rowid_reader = nullptr;
+}
+
+void ha_rocksdb::mrr_free_rows() {
+  ssize_t n_pinned = 0;
+  for (ssize_t i = 0; i < mrr_n_elements; i++) {
+    if (mrr_values[i].IsPinned()) {
+      n_pinned++;
+      mrr_values[i].Reset();
+    }
+    mrr_values[i].~PinnableSlice();
+    mrr_statuses[i].~Status();
+    // no need to free mrr_keys
+  }
+
+  // There could be rows that MultiGet has returned but MyRocks hasn't
+  // returned to the SQL layer (typically due to LIMIT clause)
+  // Count them in in "rows_read" anyway. (This is only necessary when using
+  // clustered PK. When using a secondary key, the index-only part of the scan
+  // that collects the rowids has caused all counters to be incremented)
+  if (mrr_used_cpk) stats.rows_read += n_pinned;
+
+  mrr_n_elements = 0;
+  // We can't rely on the data from HANDLER_BUFFER once the scan is over, so:
+  mrr_values = nullptr;
+}
+
+int ha_rocksdb::multi_range_read_next(char **range_info) {
+  if (mrr_uses_default_impl) {
+    return handler::multi_range_read_next(range_info);
+  }
+
+  Rdb_transaction *&tx = get_tx_from_thd(table->in_use);
+  int rc;
+
+  while (1) {
+    while (1) {
+      if (table->in_use->killed) return HA_ERR_QUERY_INTERRUPTED;
+
+      if (mrr_read_index >= mrr_n_elements) {
+        if (mrr_rowid_reader->eof() || !mrr_n_elements) {
+          table->status = STATUS_NOT_FOUND;  // not sure if this is necessary?
+          mrr_free_rows();
+          return HA_ERR_END_OF_FILE;
+        }
+
+        if ((rc = mrr_fill_buffer())) {
+          if (rc == HA_ERR_END_OF_FILE) table->status = STATUS_NOT_FOUND;
+          return rc;
+        }
+      }
+      // If we found a status that has a row, leave the loop
+      if (mrr_statuses[mrr_read_index].ok()) break;
+
+      // Skip the NotFound errors, return any other error to the SQL layer
+      if (!mrr_statuses[mrr_read_index].IsNotFound())
+        return rdb_error_to_mysql(mrr_statuses[mrr_read_index]);
+
+      mrr_read_index++;
+    }
+    size_t cur_key = mrr_read_index++;
+
+    const rocksdb::Slice &rowkey = mrr_keys[cur_key];
+
+    if (mrr_funcs.skip_record &&
+        mrr_funcs.skip_record(mrr_iter, mrr_range_ptrs[cur_key],
+                              (uchar*)rowkey.data())) {
+      rc = HA_ERR_END_OF_FILE;
+      continue;
+    }
+
+    m_last_rowkey.copy((const char *)rowkey.data(), rowkey.size(),
+                       &my_charset_bin);
+
+    *range_info = mrr_range_ptrs[cur_key];
+
+    m_retrieved_record.Reset();
+    m_retrieved_record.PinSlice(mrr_values[cur_key], &mrr_values[cur_key]);
+    mrr_values[cur_key].Reset();
```

Comment:
It looks like this was newly added, but I don't think it actually does anything.

In 
```
    m_retrieved_record.PinSlice(mrr_values[cur_key], &mrr_values[cur_key]);
```

We're calling

```
  inline void PinSlice(const Slice& s, Cleanable* cleanable) {
    assert(!pinned_);
    pinned_ = true;
    data_ = s.data();
    size_ = s.size();
    cleanable->DelegateCleanupsTo(this);
    assert(pinned_);
  }
```

and `DelegateCleanupsTo` will actually set the cleanup functions to nullptr on the origin slice. Effectively, we've already "move"d the Slice away, so `mrr_values[cur_key] ` should be empty already.

Reply:
Agree. I misread what PinnableSlice is doing. Reverted the change